### DISCUSSION TITLE
QT host

### DISCUSSION THREAD

**Discussion by redtide** [cycfi/elements#324] [Hosts]
I started to create a Qt host for elements but I'm not sure if this is really doable with the current status of Cairo Qt backend.
I made a custom QWidget where to forward all required signals/events to elements but I'm not sure how to access the host_view (mainly for the Cairo surface). Any advice?

**Comment by djowel** [comment]
I am not familiar with QT. I'm not sure how I can help. What do you mean by accessing the host_view?

**Comment by redtide** [comment]
The `host_view` class is declared in base_view.cpp, IIUC that is the one to use for cairo access to the surface.
But I see there are some problems to interface cairo with Qt, there was some work years ago but it seems it's still a work in progress.
How is going the Artist port? I saw Skia has some more documentation and examples to use with Qt.

**Comment by djowel** [comment]
Are you working on something like @jpcima's PR here: https://github.com/cycfi/elements/pull/111 ?

Artist is fine. I just need some time to integrate it. It's a bandwidth issue for me at the moment. But I'll get time one way or another.

**Comment by redtide** [comment]
Just trying to make a [Qt5 host] implementation, to be able to add elements over a view implemented in a [QWidget], inside a [QMainWindow].

[Qt5 host]: https://github.com/redtide/elements/tree/qt5-host
[QWidget]: https://doc.qt.io/qt-5/qwidget.html
[QMainWindow]: https://doc.qt.io/qt-5/qmainwindow.html

**Comment by djowel** [comment]
> Just trying to make a [Qt5 host](https://github.com/redtide/elements/tree/qt5-host) implementation, to be able to add elements over a view implemented in a [QWidget](https://doc.qt.io/qt-5/qwidget.html), inside a [QMainWindow](https://doc.qt.io/qt-5/qmainwindow.html).

That's certainly a good use-case!

**Comment by djowel** [comment]
Ok, let's see... There's actually an artist branch already (https://github.com/cycfi/elements/tree/artist_port). It's not by any means in a good state yet. I'm really tempted to have artist into elements sooner rather than later. My worry is that it will require a lot of time from me, which I do not have for the moment. The issue is not actually the port itself, but the maintenance of it after the port. The plan was to back-port element's cairo based canvas into artist first. Then have the cairo ported artist in elements as default 'canvas' backend. That would prevent a big disruption.

There's talk of a roadmap in another "issue". I'll start a draft and add this there.

In the meantime, do you have a simple example project for creating a qt window and widget I can look at, and we can discuss? I'm not a QT user, but that might inspire some ideas.

**Comment by jpcima** [comment]
If I may add to this discussion, there are multiple strategies to go for the implementation.
- create an intermediate image surface, paint on it, and transfer pixels to the qt widget
  (slow, but tolerable if UI doesn't move too much)
- extract the platform-dependent draw context (HDC/Quartz port/Xlib drawable) if possible, and make cairo paint on it
- create an OpenGL widget, setup a 2d context, and create a cairo painter to target OpenGL

**Comment by redtide** [comment]
I'm not sure what you requested is a [simple Qt example] or a basic Qt elements host case, on which I have only implemented the application and window classes, not having a view yet I can't provide a widget.

[simple Qt example]: https://github.com/redtide/qt-example
